sherpa/stats/__init__.py

```diff
@@ -474,29 +492,93 @@ def __getstate__(self):
-                  weight=None):
+                  weight=None, **kwargs):
         if not self._statfuncset:
             raise StatErr('nostat', self.name, 'calc_stat()')
-        return self.statfunc(data, model, staterror, syserror, weight)
+
+        if 'bkg' in kwargs.keys():
+            bkg = kwargs.get('bkg')
+            return self.statfunc(data, model, staterror, syserror, weight, bkg)
+        else:
+            return self.statfunc(data, model, staterror, syserror, weight)
+
+
+class WStat(Likelihood):
+    """Maximum likelihood function including background (XSPEC style).
+    This is equivalent to the X-Spec implementation of the
+    W statistic for CStat [1]_, from which the following documentation is
+    taken.
+    Suppose that each bin in the background spectrum is given its own
+    parameter so that the background model is b_i = f_i. A standard fit
+    for all these parameters would be impractical; however there is an
+    analytical solution for the best-fit f_i in terms of the other
+    variables which can be derived by using the fact that the derivative
+    of the likelihood (L) will be zero at the best fit. Solving for the
+    f_i and substituting gives the profile likelihood::
+      W = 2 sum_(i=1)^N t_s m_i + (t_s + t_b) f_i -
+          S_i ln(t_s m_i + t_s f_i) - B_i ln(t_b f_i) -
+          S_i (1- ln(S_i)) - B_i (1 - ln(B_i))
+    where
+      f_i = (S_i + B_i - (t_s + t_b) m_i + d_i) / (2 (t_s + t_b))
+    and
+      d_i = sqrt([(t_s + t_b) m_i - S_i - B_i]^2 +
+                 4(t_s + t_b) B_i m_i)
+    If any bin has S_i and/or B_i zero then its contribution to W (W_i)
+    is calculated as a special case. So, if S_i is zero then::
+      W_i = t_s m_i - B_i ln(t_b / (t_s + t_b))
+    If B_i is zero then there are two special cases. If
+    m_i < S_i / (t_s + t_b) then::
+      W_i = - t_b m_i - S_i ln(t_s / (t_s + t_b))
+    otherwise::
+      W_i = t_s m_i + S_i (ln(S_i) - ln(t_s m_i) - 1)
+    In practice, it works well for many cases but for weak sources can
+    generate an obviously wrong best fit. It is not clear why this happens
+    although binning to ensure that every bin contains at least one count
+    often seems to fix the problem. In the limit of large numbers of counts
+    per spectrum bin a second-order Taylor expansion shows that W tends to::
+      sum_(i=1)^N ( [S_i - t_s m_i - t_s f_i]^2 / (t_s (m_i + f_i)) +
+                    [B_i - t_b f_i]^2 / (t_b f_i) )
+    which is distributed as chi^2 with N - M degrees of freedom, where
+    the model m_i has M parameters (include the normalization).
+
+    References
+    ----------
+
+    .. [1] The description of the W statistic (`wstat`) in
+           https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSappendixStatistics.html
+
+    """
+
+    def __init__(self, name='wstat'):
+        Likelihood.__init__(self, name)
+
+    @staticmethod
+    def calc_stat(data, model, staterror=None, syserror=None,
```

Comment:
If I make the following changes:

    @staticmethod
    def calc_stat(data, model, datasize=None, exposuretime=None,
                  weight=None, **kwargs):

        if 'bkg' in kwargs.keys():
            bkg = kwargs.get('bkg')
        else:
            raise FitErr('no bkg file is supplied, use cstat instead')
        return _statfcts.calc_wstat_stat(data, model, datasize,
exposuretime,
                                         bkg, truncation_value)

Then it breaks symmetry with the other classes within the stats/__init__.py
file and we currently call calc_stat with 4 arguments (without keywords)
but I don't think I'm totally against this.  Hopefully no future
modifications will require the staterror and syserror keywords.

BTW, I thought about adding  two additional arguments datasize and
exposuretime to the calc_stat method of the various Statistic classes but
the problem with this approach is that it breaks backwards compatibility
with the user's existing user-stat scripts or it  makes the logical if
statement in fit.py too complicated and be error prone.


On Thu, Aug 27, 2015 at 1:55 PM, Christoph Deil <notifications@github.com>
wrote:

> In sherpa/stats/__init__.py
> <https://github.com/sherpa/sherpa/pull/94#discussion_r38125330>:
>
> > +    which is distributed as chi^2 with N - M degrees of freedom, where
> > +    the model m_i has M parameters (include the normalization).
> > +
> > +    References
> > +    ----------
> > +
> > +    .. [1] The description of the W statistic (`wstat`) in
> > +           https://heasarc.gsfc.nasa.gov/xanadu/xspec/manual/XSappendixStatistics.html
> > +
> > +    """
> > +
> > +    def __init__(self, name='wstat'):
> > +        Likelihood.__init__(self, name)
> > +
> > +    @staticmethod
> > +    def calc_stat(data, model, staterror=None, syserror=None,
>
> Thanks for implementing this!
>
> As a future user of this feature, I'd like to ask / comment:
>
> This is shoehorning wstat into a Likelihood.calc_stat interface that was
> designed for chi2 stats, right?
> I.e. arguments are re-used for something completely different than their
> name says: exposuretime = syserror
>
> I guess there is no better way to get wstat into the existing API?
> If so, can you at least explain what's going on here in the docstring?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sherpa/sherpa/pull/94/files#r38125330>.
>